### PR TITLE
feat: add ldes

### PR DESCRIPTION
**Pull request by pgallik**

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -1060,7 +1096,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        services: ['streetname-registry-api', 'streetname-registry-projections', 'streetname-registry-backoffice-api', 'streetname-registry-consumer', 'streetname-registry-producer', 'streetname-registry-migrator-streetname']
+        services: ['streetname-registry-api', 'streetname-registry-projections', 'streetname-registry-backoffice-api', 'streetname-registry-consumer', 'streetname-registry-producer', 'streetname-registry-migrator-streetname', 'streetname-registry-producer-snapshot-oslo']
```
Comment (ArneD) [inline .github/workflows/release.yml:1107]: Should be added to test: line #994

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+LABEL registry="streetname-registry"
+
+COPY / /app
+WORKDIR /app
+
+RUN apt-get update && \
+    apt-get install curl jq -y && \
+    chmod +x ./init.sh
+
+EXPOSE 4014/tcp
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Dockerfile]: wasn't it 4016?

**Review by ArneD — CHANGES_REQUESTED** [review]

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Dockerfile (hunk 2 above)
Comment (pgallik) [reply src/StreetNameRegistry.Producer.Snapshot.Oslo/Dockerfile]: yes

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Extensions/MessageExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
+namespace StreetNameRegistry.Producer.Snapshot.Oslo.Extensions
+{
+    using System.Linq;
+    using Be.Vlaanderen.Basisregisters.GrAr.Provenance;
+    using Municipality.Events;
+    using Contracts = Be.Vlaanderen.Basisregisters.GrAr.Contracts.StreetNameRegistry;
+    using ContractsCommon = Be.Vlaanderen.Basisregisters.GrAr.Contracts.Common;
+    using Domain = StreetName.Events;
+
+    public static class MessageExtensions
```
Comment (jvandaal) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Extensions/MessageExtensions.cs]: This file can go?

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/SnapshotManager.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+            var snapshotDto = DateTimeOffset.Parse(snapshot.Identificator.Versie);
+            var snapshotVersion = Instant.FromDateTimeOffset(snapshotDto);
+
+            if (eventTimestamp > snapshotVersion)
+            {
+                throw new StaleSnapshotException();
+            }
+
+            return eventTimestamp == snapshotVersion
```
Comment (jvandaal) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/SnapshotManager.cs]: Why don't we use <= ?

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                {
+                    Common =
+                    {
+                        ServiceProvider = serviceProvider,
+                        ApplicationLifetime = appLifetime
+                    }
+                });
+        }
+
+        private static string GetApiLeadingText(ApiVersionDescription description)
```
Comment (jvandaal) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/Startup.cs:190]: Text should be adjusted with the correct producer name.

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Projections/ProjectionsController.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    public class ProjectionsController : DefaultProjectorController
+    {
+        public ProjectionsController(
+            IConnectedProjectionsManager connectedProjectionsManager,
+            IConfiguration configuration)
+            : base(
+                connectedProjectionsManager,
+                configuration.GetValue<string>("BaseUrl"))
+        {
+            RegisterConnectionString(Schema.Producer, configuration.GetConnectionString("ProducerProjections"));
```
Comment (jvandaal) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Projections/ProjectionsController.cs]: Schema should change?

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Extensions/MessageExtensions.cs (hunk 3 above)
Comment (pgallik) [reply src/StreetNameRegistry.Producer.Snapshot.Oslo/Extensions/MessageExtensions.cs]: yes

**Review by jvandaal — APPROVED** [review]

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/PublicApiHttpProxy.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    using Be.Vlaanderen.Basisregisters.AspNetCore.Mvc.Formatters.Json;
+    using Be.Vlaanderen.Basisregisters.GrAr.Legacy;
+    using Newtonsoft.Json;
+
+    public interface IPublicApiHttpProxy
+    {
+        Task<OsloResult> GetSnapshot(string streetNamePersistentLocalId, CancellationToken cancellationToken);
+    }
+
+    public class PublicApiHttpProxy : IPublicApiHttpProxy
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/PublicApiHttpProxy.cs]: Let's seal this class

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/Program.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+namespace StreetNameRegistry.Producer.Snapshot.Oslo.Infrastructure
+{
+    using Be.Vlaanderen.Basisregisters.Api;
+    using Be.Vlaanderen.Basisregisters.Aws.DistributedMutex;
+    using Microsoft.AspNetCore.Hosting;
+
+    public class Program
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/Program.cs]: Seal this

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/Configuration/AddNoCacheHeadersMiddleware.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
+namespace StreetNameRegistry.Producer.Snapshot.Oslo.Infrastructure.Configuration
+{
+    using System.Threading.Tasks;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.Net.Http.Headers;
+
+    /// <summary>
+    /// Add headers to the response to prevent any caching.
+    /// </summary>
+    public class AddNoCacheHeadersMiddleware
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/Configuration/AddNoCacheHeadersMiddleware.cs]: seal this

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/SnapshotManager.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    using System.Threading.Tasks;
+    using NodaTime;
+    using Polly;
+
+    public interface ISnapshotManager
+    {
+        Task<OsloResult?> FindMatchingSnapshot(string persistentLocalId, Instant eventVersion, bool throwStaleWhenGone, CancellationToken ct);
+    }
+
+    public class SnapshotManager : ISnapshotManager
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/SnapshotManager.cs]: seal this

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/StaleSnapshotException.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+namespace StreetNameRegistry.Producer.Snapshot.Oslo.Infrastructure
+{
+    using System;
+
+    public class StaleSnapshotException : Exception
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/Infrastructure/StaleSnapshotException.cs]: Check other exceptions. Needs `[Serializable]`, sealed, ctors

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/ProducerProjections.cs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    using Be.Vlaanderen.Basisregisters.EventHandling;
+    using Be.Vlaanderen.Basisregisters.MessageHandling.Kafka.Simple;
+    using Be.Vlaanderen.Basisregisters.ProjectionHandling.Connector;
+    using Infrastructure;
+    using Microsoft.Extensions.Configuration;
+    using Municipality.Events;
+
+    [ConnectedProjectionName("Kafka producer")]
+    [ConnectedProjectionDescription("Projectie die berichten naar de kafka broker stuurt.")]
+    public class ProducerProjections : ConnectedProjection<ProducerContext>
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/ProducerProjections.cs]: seal this

File: src/StreetNameRegistry.Producer.Snapshot.Oslo/paket.references [hunk 13]
```diff
@@ -0,0 +1,16 @@
+Be.Vlaanderen.Basisregisters.Api
```
Comment (ArneD) [inline src/StreetNameRegistry.Producer.Snapshot.Oslo/paket.references:1]: can you recheck the references to see if all are needed?

File: test/StreetNameRegistry.Tests/SnapshotManagerTests.cs [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+    using System.Net;
+    using System.Net.Http;
+    using System.Threading;
+    using FluentAssertions;
+    using Moq;
+    using NodaTime;
+    using Producer.Snapshot.Oslo.Infrastructure;
+    using Xunit;
+
+    public class SnapshotManagerTests
```
Comment (ArneD) [inline test/StreetNameRegistry.Tests/SnapshotManagerTests.cs]: seal this

File: test/StreetNameRegistry.Tests/SnapshotManagerTests.cs [hunk 15]
```diff
@@ -0,0 +1,269 @@
+using System;
```
Comment (ArneD) [inline test/StreetNameRegistry.Tests/SnapshotManagerTests.cs]: usings inside namespace

**Review by ArneD — CHANGES_REQUESTED** [review]

**Pull request merged**